django/db/models/sql/datastructures.py

```diff
@@ -101,19 +106,28 @@ def as_sql(self, compiler, connection):
     def relabeled_clone(self, change_map):
         new_parent_alias = change_map.get(self.parent_alias, self.parent_alias)
         new_table_alias = change_map.get(self.table_alias, self.table_alias)
+        if self.filtered_relation is not None:
+            filtered_relation = self.filtered_relation.clone()
+            filtered_relation.path = [change_map.get(p, p) for p in self.filtered_relation.path]
+        else:
+            filtered_relation = None
         return self.__class__(
             self.table_name, new_parent_alias, new_table_alias, self.join_type,
-            self.join_field, self.nullable)
+            self.join_field, self.nullable, filtered_relation=filtered_relation)
 
-    def __eq__(self, other):
+    def equals(self, other, with_filtered_relation):
         if isinstance(other, self.__class__):
             return (
                 self.table_name == other.table_name and
                 self.parent_alias == other.parent_alias and
-                self.join_field == other.join_field
+                self.join_field == other.join_field and
+                (not with_filtered_relation or self.filtered_relation == other.filtered_relation)
             )
         return False
```

Comment:
whilst you're editing this, it can be simplified a bit - the `if isinstance(...)` can be rolled into the single `return` statement, because `and` is lazy, like you did below